pkg/network/ovn_kubernetes.go

```diff
@@ -334,3 +381,133 @@ func listenDualStack(masterIP string) string {
+	// +------------+---------------+-----------------+---------------+
+
+	// both older (than CNO)
+	// Update node only.
+	if masterDirection == directionUpgrade && nodeDirection == directionUpgrade {
+		log.Printf("Upgrading OVN-Kubernetes node before master")
+		return true, false
+	}
+	// weird: master alread upgraded, node needs upgrade
+	if masterDirection == directionUnchanged && nodeDirection == directionUpgrade {
+		return true, true
+	}
+	// weird: master needs downgrade, node needs upgrade
+	// in a bad state, just roll out changes
+	if masterDirection == directionDowngrade && nodeDirection == directionUpgrade {
+		log.Printf("WARNING: OVN-Kubernetes daemonset versions in an inconsistent state")
+		return true, true
+	}
+
+	// Next row in the chart!
+
+	// master older, node updated
+	// update master if node is rolled out
+	if masterDirection == directionUpgrade && nodeDirection == directionUnchanged {
+		if daemonSetProgressing(existingNode) {
+			log.Printf("Waiting for OVN-Kubernetes node update to roll out")
+			return true, false
+		}
+		return true, true
+	}
+
+	// unlikely, should be caught above
+	if masterDirection == directionUnchanged && nodeDirection == directionUnchanged {
+		return true, true
+	}
+
+	// weird: master needs downgrade, node already updated
+	if masterDirection == directionDowngrade && nodeDirection == directionUnchanged {
+		return true, true
+	}
+
+	// Last row
+
+	// weird: master needs downgrade, node needs upgrade
+	// in a bad state, just roll out changes
+	if masterDirection == directionUpgrade && nodeDirection == directionDowngrade {
+		log.Printf("WARNING: OVN-Kubernetes daemonset versions in an inconsistent state")
+		return true, true
+	}
+
+	// master same, node needs downgrade
+	// wait for master rollout
+	if masterDirection == directionUnchanged && nodeDirection == directionDowngrade {
+		if daemonSetProgressing(existingMaster) {
+			log.Printf("Waiting for OVN-Kubernetes master update to roll out")
+			return false, true
+		}
+		return true, true
+	}
+
+	if masterDirection == directionDowngrade && nodeDirection == directionDowngrade {
+		log.Printf("Downgrading OVN-Kubernetes master before node")
+		return false, true
+	}
+
+	// should be unreachable
+	log.Printf("WARNING: unhandled upgrade case")
+	return true, true
+}
+
+// daemonSetProgressing returns true if a daemonset is rolling out a change.
+// We also return true if the daemonset is hung at 90%, just so we don't block.
+func daemonSetProgressing(ds *appsv1.DaemonSet) bool {
+	status := ds.Status
+
+	// Copy-pasted from status_manager: Determine if a DaemonSet is progressing
+	progressing := (status.UpdatedNumberScheduled < status.DesiredNumberScheduled ||
+		status.NumberUnavailable > 0 ||
+		status.NumberAvailable == 0 ||
+		ds.Generation > status.ObservedGeneration)
+
+	if !progressing {
+		return false
+	}
+
+	// If we're hung, but max(90% of nodes, 1) have been updated, then act as if not progressing
```

Comment:
we probably want this behavior for the node DS but not the master DS?

Reply:
agreed.